src/types.ts

```diff
@@ -42,3 +43,36 @@ export type UseFetchResult<TData = any> = FetchCommands & {
 export type useFetchArg1 = string | Options & RequestInit
 
 export type UseFetch<TData> = DestructuringCommands<TData> & UseFetchResult<TData>
+
+// useFetch argument types
+
+export type UseFetchBaseOptions = {
+  onMount?: boolean,
+  timeout?: number
+}
+
+export type OptionsNoURLs = UseFetchBaseOptions & RequestInit
+
+// No Provider
+export type URLRequiredOptions = { url: string } & UseFetchBaseOptions & RequestInit
+
+// type BaseURLRequiredOptions = { baseURL: string } & UseFetchBaseOptions & RequestInit
+
+export type OptionsAsFirstParam = URLRequiredOptions// | BaseURLRequiredOptions
+
+// With Provider
+export type MaybeURLOptions = { url?: string } & UseFetchBaseOptions & RequestInit
+
+// type MaybeBaseURLOptions = { baseURL?: string } & UseFetchBaseOptions & RequestInit
+
+export type MaybeOptions = MaybeURLOptions //| MaybeBaseURLOptions
```

Comment:
Why have an alias, why not just use `MaybeURLOptions`

Reply:
so, the reason is that I'm trying to figure out all the cases. Do we remove `baseURL` or keep it. In this case, I'm assuming we can remove it. Btw, I would totally be down to pair with you on some of this if you're interested :)

Reply:
yes, definitely up for pairing although we'll need to find a convenient time for our respective time zones

Reply:
I am in the UK

Reply:
I think this is all seems very busy for what is trying to be achieved.  I'm sure this all can be simplified.  Might need to take a step back and first of all get a simpler list of types before worrying about splitting them up.  I think there is too much happening here.   Maybe the best thing to do is get good test coverage and then mercylessly refactor until there is a clean API.

Reply:
I'm gong to list the types in a separate branch to try and understand this better.

Reply:
I think what you should be striving for is something like this.  Please feel free to disagree with any of this

```javascript
import { UseFetch } from "./types";

export type UseFetchOptions = RequestInit & {
  url?: string;
  onMount: boolean;
  timeout: number;
}

function useFetch<TData = any>(url: string, options?: Omit<UseFetchOptions, 'url'>): UseFetch<TData>
function useFetch<TData = any>(options: UseFetchOptions): UseFetch<TData>
function useFetch<TData = any>(urlOrOptions: string | UseFetchOptions): UseFetch<TData> {}
```

Reply:
You are probably right, I could be overthinking this. Why don't we get this out since it's working, and then we can come back and refactor again. It's just 1. there's fixes in this 2. I don't wanna block tests and 3. don't wanna block your `retry` PR

Reply:
> I think what you should be striving for is something like this.
> 
> I don't think there really needs to be any more but you will know better than me.
> 
> Please feel free to disagree with any of this
> 
> ```js
> import { UseFetch } from "./types";
> 
> export type UseFetchOptions = RequestInit & {
>   url?: string;
>   onMount: boolean;
>   timeout: number;
> }
> 
> function useFetch<TData = any>(url: string, options?: Omit<UseFetchOptions, 'url'>): UseFetch<TData>
> function useFetch<TData = any>(options: UseFetchOptions): UseFetch<TData>
> function useFetch<TData = any>(urlOrOptions: string | UseFetchOptions): UseFetch<TData> {}
> ```

So, `onMount` and `timeout` are also optional 🙂And this looks awesome, I will change it to be like this. Thanks!

Reply:
@alex-cory I think the overloaded function would actually look like this:

```javascript
export function useFetch<TData = any, TReturnType = void>(url: string, options?: Omit<UseFetchOptions2, 'url'>): UseFetch<TData>
export function useFetch<TData = any, TReturnType = void>(fullOptions: UseFetchOptions2): UseFetch<TData>
export function useFetch<TData = any, TReturnType = void>(urlOrOptions: string | UseFetchOptions2, potentialOptions?: Omit<UseFetchOptions2, 'url'>): UseFetch<TData> {
```

The `TReturnType` would be the type used in `Promise`

Reply:
> @alex-cory I think the overloaded function would actually look like this:
> 
> ```js
> export function useFetch<TData = any, TReturnType = void>(url: string, options?: Omit<UseFetchOptions2, 'url'>): UseFetch<TData>
> export function useFetch<TData = any, TReturnType = void>(fullOptions: UseFetchOptions2): UseFetch<TData>
> export function useFetch<TData = any, TReturnType = void>(urlOrOptions: string | UseFetchOptions2, potentialOptions?: Omit<UseFetchOptions2, 'url'>): UseFetch<TData> {
> ```
> 
> The `TReturnType` would be the type used in `Promise`

What do you think about the changes I just made. I didn't see what you had until I just pushed this.